File: algorithms.py
"""
Mathematical Algorithms and Calculations (Python Version)
Core algorithms for task scheduling, energy analysis, and productivity metrics

This Python module can be used for backend processing, data analysis, or ML training.
"""

from typing import List, Dict, Tuple
from dataclasses import dataclass


@dataclass
class EnergyPattern:
    """Energy pattern data point"""
    hour: int
    energy: float  # 0-100


# Default circadian energy pattern based on human biology
# Peak energy typically occurs mid-morning (9-11 AM)
DEFAULT_ENERGY_PATTERN: List[EnergyPattern] = [
    EnergyPattern(6, 30),
    EnergyPattern(7, 50),
    EnergyPattern(8, 70),
    EnergyPattern(9, 85),
    EnergyPattern(10, 90),  # Peak energy
    EnergyPattern(11, 85),
    EnergyPattern(12, 75),
    EnergyPattern(13, 60),  # Post-lunch dip
    EnergyPattern(14, 50),
    EnergyPattern(15, 55),
    EnergyPattern(16, 70),
    EnergyPattern(17, 75),
    EnergyPattern(18, 70),
    EnergyPattern(19, 60),
    EnergyPattern(20, 50),
    EnergyPattern(21, 40),
    EnergyPattern(22, 30),
    EnergyPattern(23, 20),
]


def find_optimal_time_windows(
    energy_required: float,
    energy_pattern: List[EnergyPattern] = None
) -> List[int]:
    """
    Calculate optimal task scheduling time based on energy requirements.
    
    Args:
        energy_required: Energy level needed for task (0-100)
        energy_pattern: Energy pattern data (defaults to DEFAULT_ENERGY_PATTERN)
    
    Returns:
        Array of optimal time windows (hours) for the task
    """
    if energy_pattern is None:
        energy_pattern = DEFAULT_ENERGY_PATTERN
    
    optimal_hours = []
    for pattern in energy_pattern:
        if pattern.energy >= energy_required:
            optimal_hours.append(pattern.hour)
    
    return optimal_hours


def calculate_peak_energy(
    energy_pattern: List[EnergyPattern] = None
) -> Dict[str, float]:
    """
    Calculate peak energy window.
    
    Args:
        energy_pattern: Energy pattern data (defaults to DEFAULT_ENERGY_PATTERN)
    
    Returns:
        Dictionary with peak hour and energy level
    """
    if energy_pattern is None:
        energy_pattern = DEFAULT_ENERGY_PATTERN
    
    peak = max(energy_pattern, key=lambda x: x.energy)
    return {"hour": peak.hour, "energy": peak.energy}


def calculate_productivity_score(
    completed_tasks: int,
    total_tasks: int,
    energy_alignment: float = 0.5
) -> float:
    """
    Calculate productivity score based on completed tasks and energy alignment.
    
    Args:
        completed_tasks: Number of completed tasks
        total_tasks: Total number of tasks
        energy_alignment: How well tasks were aligned with energy levels (0-1)
    
    Returns:
        Productivity score (0-100)
    """
    if total_tasks == 0:
        return 0.0
    
    completion_rate = completed_tasks / total_tasks
    base_score = completion_rate * 100
    alignment_bonus = energy_alignment * 20  # Up to 20 points bonus
    
    return min(100.0, round(base_score + alignment_bonus, 1))


def calculate_health_score(
    sleep_hours: float,
    break_count: int,
    work_hours: float
) -> float:
    """
    Calculate health score based on sleep, breaks, and work-life balance.
    
    Args:
        sleep_hours: Hours of sleep
        break_count: Number of breaks taken
        work_hours: Total work hours
    
    Returns:
        Health score (0-100)
    """
    # Sleep score (optimal: 7-9 hours)
    if 7 <= sleep_hours <= 9:
        sleep_score = 40.0  # Optimal sleep
    elif 6 <= sleep_hours <= 10:
        sleep_score = 30.0  # Acceptable
    else:
        sleep_score = max(0.0, 40.0 - abs(sleep_hours - 8) * 5)
    
    # Break score (optimal: break every 90 minutes)
    optimal_breaks = int(work_hours / 1.5)
    break_score = min(30.0, (break_count / max(optimal_breaks, 1)) * 30)
    
    # Work-life balance (optimal: 8 hours work)
    if work_hours <= 8:
        balance_score = 30.0
    else:
        balance_score = max(0.0, 30.0 - (work_hours - 8) * 3)
    
    return min(100.0, round(sleep_score + break_score + balance_score, 1))


def calculate_duration(start_time: str, end_time: str) -> int:
    """
    Calculate task duration in minutes from two time strings.
    
    Args:
        start_time: Start time string (HH:MM format)
        end_time: End time string (HH:MM format)
    
    Returns:
        Duration in minutes
    """
    from datetime import datetime, timedelta
    
    start = datetime.strptime(start_time, "%H:%M")
    end = datetime.strptime(end_time, "%H:%M")
    
    # Handle overnight events
    if end < start:
        end += timedelta(days=1)
    
    delta = end - start
    return int(delta.total_seconds() / 60)


def find_deep_work_windows(
    energy_pattern: List[EnergyPattern] = None
) -> List[int]:
    """
    Find deep work windows (energy >= 70).
    
    Args:
        energy_pattern: Energy pattern data (defaults to DEFAULT_ENERGY_PATTERN)
    
    Returns:
        Array of hours suitable for deep work
    """
    if energy_pattern is None:
        energy_pattern = DEFAULT_ENERGY_PATTERN
    
    return [pattern.hour for pattern in energy_pattern if pattern.energy >= 70]


def calculate_energy_alignment(
    scheduled_hour: int,
    task_energy_required: float,
    energy_pattern: List[EnergyPattern] = None
) -> float:
    """
    Calculate energy alignment score for a task scheduled at a specific hour.
    
    Args:
        scheduled_hour: Hour when task is scheduled (0-23)
        task_energy_required: Energy required for the task (0-100)
        energy_pattern: Energy pattern data (defaults to DEFAULT_ENERGY_PATTERN)
    
    Returns:
        Alignment score (0-1), where 1 is perfect alignment
    """
    if energy_pattern is None:
        energy_pattern = DEFAULT_ENERGY_PATTERN
    
    hour_data = next((p for p in energy_pattern if p.hour == scheduled_hour), None)
    if not hour_data:
        return 0.0
    
    available_energy = hour_data.energy
    
    # Perfect alignment: available energy matches required energy
    if available_energy >= task_energy_required:
        # Bonus if we have extra energy (but not too much)
        excess = available_energy - task_energy_required
        return min(1.0, 1.0 - (excess / 100))
    else:
        # Penalty for insufficient energy
        deficit = task_energy_required - available_energy
        return max(0.0, 1.0 - (deficit / 50))


if __name__ == "__main__":
    # Example usage
    print("Energy Pattern Algorithms")
    print("=" * 40)
    
    peak = calculate_peak_energy()
    print(f"Peak Energy: Hour {peak['hour']}, Level {peak['energy']}")
    
    deep_work = find_deep_work_windows()
    print(f"Deep Work Windows: {deep_work}")
    
    optimal = find_optimal_time_windows(80)
    print(f"Optimal times for high-energy tasks (80+): {optimal}")

File: algorithms.ts
/**
 * Mathematical Algorithms and Calculations
 * Core algorithms for task scheduling, energy analysis, and productivity metrics
 */

export interface EnergyPattern {
  hour: number;
  energy: number; // 0-100
}

/**
 * Default circadian energy pattern based on human biology
 * Peak energy typically occurs mid-morning (9-11 AM)
 */
export const DEFAULT_ENERGY_PATTERN: EnergyPattern[] = [
  { hour: 6, energy: 30 },
  { hour: 7, energy: 50 },
  { hour: 8, energy: 70 },
  { hour: 9, energy: 85 },
  { hour: 10, energy: 90 }, // Peak energy
  { hour: 11, energy: 85 },
  { hour: 12, energy: 75 },
  { hour: 13, energy: 60 }, // Post-lunch dip
  { hour: 14, energy: 50 },
  { hour: 15, energy: 55 },
  { hour: 16, energy: 70 },
  { hour: 17, energy: 75 },
  { hour: 18, energy: 70 },
  { hour: 19, energy: 60 },
  { hour: 20, energy: 50 },
  { hour: 21, energy: 40 },
  { hour: 22, energy: 30 },
  { hour: 23, energy: 20 },
];

/**
 * Calculate optimal task scheduling time based on energy requirements
 * @param energyRequired - Energy level needed for task (0-100)
 * @param energyPattern - Energy pattern data
 * @returns Array of optimal time windows (hours) for the task
 */
export function findOptimalTimeWindows(
  energyRequired: number,
  energyPattern: EnergyPattern[] = DEFAULT_ENERGY_PATTERN
): number[] {
  const optimalHours: number[] = [];
  
  energyPattern.forEach(({ hour, energy }) => {
    if (energy >= energyRequired) {
      optimalHours.push(hour);
    }
  });
  
  return optimalHours;
}

/**
 * Calculate peak energy window
 * @param energyPattern - Energy pattern data
 * @returns Object with peak hour and energy level
 */
export function calculatePeakEnergy(
  energyPattern: EnergyPattern[] = DEFAULT_ENERGY_PATTERN
): { hour: number; energy: number } {
  const peak = energyPattern.reduce((max, current) => 
    current.energy > max.energy ? current : max
  );
  
  return { hour: peak.hour, energy: peak.energy };
}

/**
 * Calculate productivity score based on completed tasks and energy alignment
 * @param completedTasks - Number of completed tasks
 * @param totalTasks - Total number of tasks
 * @param energyAlignment - How well tasks were aligned with energy levels (0-1)
 * @returns Productivity score (0-100)
 */
export function calculateProductivityScore(
  completedTasks: number,
  totalTasks: number,
  energyAlignment: number = 0.5
): number {
  const completionRate = totalTasks > 0 ? completedTasks / totalTasks : 0;
  const baseScore = completionRate * 100;
  const alignmentBonus = energyAlignment * 20; // Up to 20 points bonus
  
  return Math.min(100, Math.round(baseScore + alignmentBonus));
}

/**
 * Calculate health score based on sleep, breaks, and work-life balance
 * @param sleepHours - Hours of sleep
 * @param breakCount - Number of breaks taken
 * @param workHours - Total work hours
 * @returns Health score (0-100)
 */
export function calculateHealthScore(
  sleepHours: number,
  breakCount: number,
  workHours: number
): number {
  // Sleep score (optimal: 7-9 hours)
  let sleepScore = 0;
  if (sleepHours >= 7 && sleepHours <= 9) {
    sleepScore = 40; // Optimal sleep
  } else if (sleepHours >= 6 && sleepHours <= 10) {
    sleepScore = 30; // Acceptable
  } else {
    sleepScore = Math.max(0, 40 - Math.abs(sleepHours - 8) * 5);
  }
  
  // Break score (optimal: break every 90 minutes)
  const optimalBreaks = Math.floor(workHours / 1.5);
  const breakScore = Math.min(30, (breakCount / Math.max(optimalBreaks, 1)) * 30);
  
  // Work-life balance (optimal: 8 hours work)
  const balanceScore = workHours <= 8 ? 30 : Math.max(0, 30 - (workHours - 8) * 3);
  
  return Math.min(100, Math.round(sleepScore + breakScore + balanceScore));
}

/**
 * Calculate task duration in minutes from two time strings
 * @param startTime - Start time string (HH:MM format)
 * @param endTime - End time string (HH:MM format)
 * @returns Duration in minutes
 */
export function calculateDuration(startTime: string, endTime: string): number {
  const [startHours, startMinutes] = startTime.split(':').map(Number);
  const [endHours, endMinutes] = endTime.split(':').map(Number);
  
  const start = new Date();
  start.setHours(startHours, startMinutes, 0, 0);
  
  const end = new Date();
  end.setHours(endHours, endMinutes, 0, 0);
  
  // Handle overnight events
  if (end < start) {
    end.setDate(end.getDate() + 1);
  }
  
  return Math.round((end.getTime() - start.getTime()) / 60000);
}

/**
 * Calculate total estimated time for pending tasks
 * @param tasks - Array of tasks with estimated_duration
 * @returns Total time in minutes
 */
export function calculateTotalTaskTime(tasks: Array<{ estimated_duration: number; status: string }>): number {
  return tasks
    .filter(task => task.status !== 'completed')
    .reduce((total, task) => total + task.estimated_duration, 0);
}

/**
 * Find deep work windows (energy >= 70)
 * @param energyPattern - Energy pattern data
 * @returns Array of hours suitable for deep work
 */
export function findDeepWorkWindows(
  energyPattern: EnergyPattern[] = DEFAULT_ENERGY_PATTERN
): number[] {
  return energyPattern
    .filter(({ energy }) => energy >= 70)
    .map(({ hour }) => hour);
}

/**
 * Calculate energy alignment score for a task scheduled at a specific hour
 * @param scheduledHour - Hour when task is scheduled (0-23)
 * @param taskEnergyRequired - Energy required for the task (0-100)
 * @param energyPattern - Energy pattern data
 * @returns Alignment score (0-1), where 1 is perfect alignment
 */
export function calculateEnergyAlignment(
  scheduledHour: number,
  taskEnergyRequired: number,
  energyPattern: EnergyPattern[] = DEFAULT_ENERGY_PATTERN
): number {
  const hourData = energyPattern.find(({ hour }) => hour === scheduledHour);
  if (!hourData) return 0;
  
  const availableEnergy = hourData.energy;
  
  // Perfect alignment: available energy matches required energy
  if (availableEnergy >= taskEnergyRequired) {
    // Bonus if we have extra energy (but not too much)
    const excess = availableEnergy - taskEnergyRequired;
    return Math.min(1, 1 - (excess / 100));
  } else {
    // Penalty for insufficient energy
    const deficit = taskEnergyRequired - availableEnergy;
    return Math.max(0, 1 - (deficit / 50));
  }
}

